### DISCUSSION TITLE
DataArray.sortby() does not work as documented

### DISCUSSION THREAD

**Discussion by jinqj** [pydata/xarray#6009] [Q&A]
Hello,

I read a lot about questions and solutions related to xarray in this discussion forum. They are super useful and saved me time to dig out the issue and find fixes by myself. 

This is my first question posted here and I hope someone could help me understand how the sortby() method should be used.

Recently I found something weird about DataArray.sortby() method. Here is the code:

```python
import numpy  as np
import pandas as pd
import xarray as xr

data = np.arange(30).reshape((3,10))
data[1, 1] = 0

da = xr.DataArray(data, dims=('lat', 'lon'))

print(data)
print(da)

print(da.sortby('lon', ascending=True))
print(da.sortby('lon', ascending=False))

print(da.sortby('lat', ascending=True))
print(da.sortby('lat', ascending=False))
```

Here is the print out:

```
[[ 0  1  2  3  4  5  6  7  8  9]
 [10  0 12 13 14 15 16 17 18 19]
 [20 21 22 23 24 25 26 27 28 29]]

<xarray.DataArray (lat: 3, lon: 10)>
array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10,  0, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29]])
Dimensions without coordinates: lat, lon

<xarray.DataArray (lat: 3, lon: 10)>
array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10,  0, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29]])
Dimensions without coordinates: lat, lon

<xarray.DataArray (lat: 3, lon: 10)>
array([[ 9,  8,  7,  6,  5,  4,  3,  2,  1,  0],
       [19, 18, 17, 16, 15, 14, 13, 12,  0, 10],
       [29, 28, 27, 26, 25, 24, 23, 22, 21, 20]])
Dimensions without coordinates: lat, lon

<xarray.DataArray (lat: 3, lon: 10)>
array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10,  0, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29]])
Dimensions without coordinates: lat, lon

<xarray.DataArray (lat: 3, lon: 10)>
array([[20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [10,  0, 12, 13, 14, 15, 16, 17, 18, 19],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9]])
Dimensions without coordinates: lat, lon

```
My questions are:

1) Does sortby() accept dimensions without coordinates as its parameter of "variables"?
    Based on the documentation of xarray, it seems NO. "variables" should be "1D DataArray objects or name(s) of **1D variable(s) in coords** whose values are used to sort this array."
    I tested "lat" and "lon" dimensions that are without coordinate, but sortby() does not raise an Error instead it still does something.

2) "da.sortby('lon', ascending=True)" returns the same values as da.values, which is reasonable. However, "da.sortby('lon', ascending=False)" returns the array with column values in a reversed order. Note that it does not sort the column values but just reversed the order ([19, 18, 17, 16, 15, 14, 13, 12,  **0**, 10]. Such behavior is held for the row values "lat".

I am using xarray with the version of
`xarray                    0.19.0             pyhd8ed1ab_1    conda-forge`

Is this how sortby() supposed to work or did I get something messed up?

Thanks for your time and look forward to a discussion.
Qinjian Jin

**Comment by mathause** [comment]
You seem to have answered (1) yourself. It does work - what happens is that non-coord dimensions have implicit coords which just enumerate from 0 to len(dim). Try:

```python
da["lat"]
```

In (2) it does not sort the values but sorts _by_ the "coordinates" - as the coords are 0..9 it reverses them. Two more examples:

- Sorting arrays with coords makes more sense:
```python
da2 = xr.DataArray([1, 2, 3], coords=dict(lon=[2, 1, 3]))
da2.sortby('lon')
```

- Calling `.sortby('lon', ascending=False)` twice you see a different behavior if there coords are available:

```python
# the second sortby reverses again
da.sortby('lon', ascending=False).sortby('lon', ascending=False)

da1 = xr.DataArray([1, 2, 3], coords=dict(lon=[1, 2, 3]))
# the second sortby has no effect
da1.sortby('lon', ascending=False).sortby('lon', ascending=False)
```

**Reply by jinqj** [reply]
Hello Mathause,

Thanks a lot for your detailed answers! The example (`calling .sorby('lon', ascending=False`) is so interesting. I just wanna make sure that my understanding about the this example is correct.

- When coords are available, the results from calling `.sortby` twice is easy to understand.

- When coords are not available, each calling of `sortby` makes a difference, because the implicit coords associated with the returned array from the first calling of sortby is still 0..len(dim). That's why the second calling reverses the array again. In other words, the implicit coords are always 0..len(dim) no matter what operations has been performed on the array.

I'd like to say the behavior of xarray is so interesting and there is really a lot I need to learn. 

Last question, do you know where I can find the explanation about the hidden/implicit coords of xarray on its [official website](http://xarray.pydata.org/en/stable/)?

Thank you again,
Qinjian

**Comment by mathause** [comment]
Yes your understanding is correct. They are called [Non-dimension-coordinates](https://xarray.pydata.org/en/stable/user-guide/terminology.html#term-Non-dimension-coordinate) but I am not sure if there is much more info in the docs.

**Reply by jinqj** [reply]
It makes a lot of sense. Thank you so much for your detailed explanation.
